merging: add metrics for shard count and janitor jobs

stefanhengl opened the pull request:
This change gives us real time metrics for compound shard count, vacuuming and merging.

Co-authored-by: geoffrey@sourcegraph.com

stefanhengl committed:
merging: add metrics for shard count and janitor jobs

This change gives us real time metrics for shard count and boolean
metrics for vacuum and merging.

Co-authored-by: geoffrey@sourcegraph.com

tsenart commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -673,6 +673,25 @@ func getEnvWithDefaultInt64(k string, defaultVal int64) int64 {
+		return
+	}
+	defer d.Close()
+	fns, _ := d.Readdirnames(-1)
```
nit: What does `fns` stand for here?

tsenart replied:
I think we could use https://pkg.go.dev/io/fs#Glob here.

stefanhengl replied:
`fns` stands for filenames, we use it in cleanup.go, too.

stefanhengl replied:
Glob sounds good. I chose filepath.Glob over fs.Glob because we already use it in watcher.go and other places. Any reason to prefer fs.Glob here?

stefanhengl committed:
use glob instead of Readdirnames

stefanhengl committed:
track merge duration and number of errors

ggilmore commented on cmd/zoekt-sourcegraph-indexserver/merge.go:
```diff
@@ -47,12 +67,16 @@ func doMerge(dir string, targetSizeBytes, maxSizeBytes int64, simulate bool) err
 		if !simulate {
+			start := time.Now()
 			stdOut, stdErr, err := callMerge(comp.shards)
+			metricShardMergingDuration.Observe(time.Since(start).Seconds())
```
Not blocking on this, but ideally we'd want to have a _separate_ metric for the duration of failed merging operations

stefanhengl replied:
We don't really expect that many failed merges. It should be a rare event. According to our log, there were 0 errors during merging over last 30 days. I think the error counter already gives us a good enough signal for now.

tsenart replied:
Even if we don't have failures so far (great!), I think we should standardize on always having a normalized `error` label in latency histograms.

stefanhengl replied:
done

stefanhengl committed:
improve docstrings and buckets

tsenart commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -673,6 +673,23 @@ func getEnvWithDefaultInt64(k string, defaultVal int64) int64 {
 }
 
+func initializeCompoundShardCounter(indexDir string) {
+	fns, err := filepath.Glob(filepath.Join(indexDir, "*.zoekt"))
```
I was reading the code and I don't think there's a big difference between io/fs.Glob and path/filepath.Glob. The former was introduced more recently, so it's the `io.FS` aware version of Glob, with the extra argument.

I think you can do the filtering of the compound- prefix here already.

```suggestion
	fns, err := filepath.Glob(filepath.Join(indexDir, "compound-*.zoekt"))
```

tsenart replied:
And then there's no need to loop over the fns: `metricNumberCompoundShards.Set(float64(len(fns)))`

tsenart reviewed (approved):
Left some more feedback, approving to unblock.

stefanhengl committed:
add label "error" for tracking duration

stefanhengl committed:
address PR comments

stefanhengl committed:
remove error counter

we are tracking this with the label "error" in
mertricShardMergingDuration instead.
